### PR TITLE
incr.comp.: Turn translation-related attributes into a query

### LINKED ISSUE #47320: incr.comp.: Turn translation-related attributes into a query.
There are a few attributes, like `#[inline]`, `#[cold]`, and `#[target_feature]`, that would profit from being turned into a query:
- directly accessing the `ast::Attribute` will lead to a dependency edge to HIR which is brittle because of spans,
- at least #[inline] is searched for a few times, so caching might be good.

The attributes in question are listed in `trans::attributes::from_fn_attrs()`: https://github.com/rust-lang/rust/blob/27ede55414e01f13c6869a8763da207e544cc6ad/src/librustc_trans/attributes.rs#L97-L132

A query could return a `TransFnAttrs` value that looks like:
```rust
struct TransFnAttrs {
    inline: InlineAttr, // from syntax::attr
    flags: TransFnAttrFlags,
    target_features: Symbol,
}

bitflags! {
    pub struct TransFnAttrFlags: u8 {
        const COLD                      = 0b0000_0001;
        const ALLOCATOR                 = 0b0000_0010;
        const UNWIND                    = 0b0000_0100;
        const RUSTC_ALLOCATOR_NOUNWIND  = 0b0000_1000;
        const NAKED                     = 0b0001_0000;
    }
}
```

The `syntax::attr::{find_inline_attr, requests_inline}` should be replaced by this new query then. 

The ultimate goal of this refactoring is to reduce the number of false positives during incr. comp. cache invalidation by providing a "firewall" between HIR and the `compile_codegen_unit` query.

### PR DESCRIPTION
**Pull request by wesleywiser**
Fixes #47320

r? @michaelwoerister

### CODE REVIEW

File: src/librustc_mir/transform/inline.rs [hunk 1]
```diff
@@ -207,7 +207,7 @@ impl<'a, 'tcx> Inliner<'a, 'tcx> {
         }
 
         let attrs = tcx.get_attrs(callsite.callee);
```
Comment (michaelwoerister) [inline src/librustc_mir/transform/inline.rs]: We should be able to get rid of this `attrs` entirely.

Comment (michaelwoerister) [reply src/librustc_mir/transform/inline.rs]: Just saw that this is fixed in a later commit `:)`

File: src/librustc_trans_utils/trans_crate.rs [hunk 2]
```diff
@@ -198,8 +199,9 @@ impl TransCrate for MetadataOnlyTransCrate {
 
     fn provide(&self, providers: &mut Providers) {
         ::symbol_names::provide(providers);
-        providers.target_features_enabled = |_tcx, _id| {
-            Lrc::new(Vec::new()) // Just a dummy
+
+        providers.target_features_whitelist = |_tcx, _cnum| {
+            Rc::new(FxHashSet()) // Just a dummy
```
Comment (michaelwoerister) [inline src/librustc_trans_utils/trans_crate.rs]: This will probably need updating to `Lrc` too.

**Review by michaelwoerister — COMMENTED** [review]
Thanks a lot for the very thorough PR, @wesleywiser! This makes a lot of things much nicer, I think, and probably more efficient too.

I added some comments where things need addressing. Just minor things though.

One thing left to do is caching the new query. Doing so is a bit more involved than it should be but it's pretty simple nonetheless:
1. Mark the query as cacheable like here: https://github.com/rust-lang/rust/blob/e2746d870017c869b84fc56ecd63f5e4f9646c96/src/librustc/ty/maps/config.rs#L685
2. Allow for the query to be "refreshed" before writing back out to disk: https://github.com/rust-lang/rust/blob/e2746d870017c869b84fc56ecd63f5e4f9646c96/src/librustc/ty/maps/plumbing.rs#L999
3. Actually write query results to disk, like here: https://github.com/rust-lang/rust/blob/e2746d870017c869b84fc56ecd63f5e4f9646c96/src/librustc/ty/maps/on_disk_cache.rs#L220

File: src/librustc/ich/impls_hir.rs [hunk 3]
```diff
@@ -1156,6 +1158,14 @@ impl<'hir> HashStable<StableHashingContext<'hir>> for hir::TransFnAttrFlags
     }
 }
 
+impl<'hir> HashStable<StableHashingContext<'hir>> for attr::InlineAttr {
```
[the remark this answers is not in the document]
Comment (wesleywiser) [reply src/librustc/ich/impls_hir.rs]: I tried that but the macro doesn't seem to like the path to that enum. Trying `impl_stable_hash_for!(enum attr::InlineAttr { ... })` yields `error[E0433]: failed to resolve. Did you mean syntax::attr?`

but changing that to `impl_stable_hash_for!(enum syntax::attr::InlineAttr { ... })` yields `error[E0433]: failed to resolve. Use of undeclared type or module syntax`. I'm not sure where to go from here...

**Comment by wesleywiser** [comment]
@michaelwoerister I fixed all of your feedback except for the issue I mentioned above with the `impl_stable_hash_for!(enum)` macro.

**Comment by michaelwoerister** [comment]
Yes, the macro is sometimes weird like that. With an absolute path (starting with `::syntax`) it would probably work.

Thanks for fixing the nits! This looks great!

@bors try (I'd like to see if this affects performance)

@bors r+

**Comment by michaelwoerister** [comment]
I seems that bors has decided to push this PR to the front of the queue (and kill the one that was almost finished `:(`). If it goes through, we don't need a separate perf run. Note to myself: don't `r+` and `try` at the same time...

**Comment by michaelwoerister** [comment]
@Mark-Simulacrum, could you do a perf run please?

**Comment by Mark-Simulacrum** [comment]
Perf started for b773dbdc99823e59733116d290342079ba1ca26e.

**Comment by michaelwoerister** [comment]
Thanks, @Mark-Simulacrum!

Here's the link: http://perf.rust-lang.org/compare.html?start=1733a61141d125beb45587dd89d54cd4a01cdd5a&end=b773dbdc99823e59733116d290342079ba1ca26e&stat=instructions%3Au

Will this work? Are there build artifacts for the commit? The try-build was f24577a.

**Comment by wesleywiser** [comment]
Rebased

**Comment by Mark-Simulacrum** [comment]
There were two try builds, of which bors seems to think only one completed so I ran perf for that one. The URL you provided looks correct.

**Comment by michaelwoerister** [comment]
Thanks, @Mark-Simulacrum, yes the link works now, it's all good `:)`

The numbers look good too! Pretty much in line with what I expected.

**Comment by wesleywiser** [comment]
@michaelwoerister Are you concerned about the `style-servo` and `style-servo-opt` max regressions?

**Comment by michaelwoerister** [comment]
No, that's fine, the clean-incremental benchmark for style-servo is not stable: https://github.com/rust-lang/rust/issues/48184 
There's something in there that causes a varying number of CGUs to be recompiled even if there is no change.

**Comment by wesleywiser** [comment]
Got it. Thanks!

**Pull request merged**